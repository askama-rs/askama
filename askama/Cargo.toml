[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
