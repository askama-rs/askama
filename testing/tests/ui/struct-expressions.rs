#![allow(unused_variables)]

use askama::Template;

#[derive(Default)]
pub struct InnerCell {
    pub number: u32,
    pub other: u32,
}
impl std::fmt::Display for InnerCell {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Hello, {} and {}!", self.number, self.other)
    }
}

#[derive(Template)]
#[template(
    source = r#"{{ InnerCell { ..Default::default(), other: 24 } }}"#,
    ext = "txt"
)]
struct OuterTemplate1;

#[derive(Template)]
#[template(source = r#"some{{ InnerCell { }}thing"#, ext = "txt")]
struct OuterTemplate2;

#[derive(Template)]
#[template(source = r#"some{{ InnerCell { foo }}thing"#, ext = "txt")]
struct OuterTemplate3;

#[derive(Template)]
#[template(source = r#"some{{ InnerCell { foo: }}thing"#, ext = "txt")]
struct OuterTemplate4;

#[derive(Template)]
#[template(source = r#"some{{ InnerCell { foo: 32 }}thing"#, ext = "txt")]
struct OuterTemplate5;

#[derive(Template)]
#[template(source = r#"some{{ InnerCell { foo: 32, }}thing"#, ext = "txt")]
struct OuterTemplate6;

#[derive(Template)]
#[template(source = r#"some{{ InnerCell { foo: 32, .. }}thing"#, ext = "txt")]
struct OuterTemplate7;

#[derive(Template)]
#[template(source = r#"some{{ InnerCell { foo: 32, ..Default::Default }}thing"#, ext = "txt")]
struct OuterTemplate8;

fn main() {}
