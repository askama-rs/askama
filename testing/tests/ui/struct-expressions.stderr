error: expected `}` after `..` was used
 --> <source attribute>:1:37
       "other: 24 } }}"
  --> tests/ui/struct-expressions.rs:19:14
   |
19 |     source = r#"{{ InnerCell { ..Default::default(), other: 24 } }}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: failed to parse template source
 --> <source attribute>:1:20
       "}"
  --> tests/ui/struct-expressions.rs:25:21
   |
25 | #[template(source = r#"some{{ InnerCell { }}"#, ext = "txt")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: missing `}` at the end of the struct
 --> <source attribute>:1:17
       "{ foo: } }}"
  --> tests/ui/struct-expressions.rs:29:21
   |
29 | #[template(source = r#"some{{ InnerCell { foo: } }}"#, ext = "txt")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: missing `}` at the end of the struct
 --> <source attribute>:1:17
       "{ foo: 32, .. } }}"
  --> tests/ui/struct-expressions.rs:33:21
   |
33 | #[template(source = r#"some{{ InnerCell { foo: 32, .. } }}"#, ext = "txt")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: missing `}` at the end of the struct
 --> <source attribute>:1:17
       "{ . } }}"
  --> tests/ui/struct-expressions.rs:37:21
   |
37 | #[template(source = r#"some{{ InnerCell { . } }}"#, ext = "txt")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: missing `}` at the end of the struct
 --> <source attribute>:1:17
       "{ , } }}"
  --> tests/ui/struct-expressions.rs:41:21
   |
41 | #[template(source = r#"some{{ InnerCell { , } }}"#, ext = "txt")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: missing `}` at the end of the struct
 --> <source attribute>:1:17
       "{ ..Default::default(): 32 } }}"
  --> tests/ui/struct-expressions.rs:45:21
   |
45 | #[template(source = r#"some{{ InnerCell { ..Default::default(): 32 } }}"#, ext = "txt")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: didn't expect arguments in base call
 --> <source attribute>:1:19
       "..Default::default(32) } }}"
  --> tests/ui/struct-expressions.rs:49:21
   |
49 | #[template(source = r#"some{{ InnerCell { ..Default::default(32) } }}"#, ext = "txt")]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
